---
title: "Running the analysis on other mutation data"
author: "Van den Eynden Jimmy"
date: "02/04/2019"
output: html_document
---

This document shortly explains how to 1) apply the HLA-binding annotation and 2) calculate HBMR and dNnonHLA/dNHLA ratios using your own mutation data. 
It will use the TCGA MC3 calls as an example but can be easily adapted to any other dataset.

# Download example dataset

We will use the TCGA MC3 calls as reported by *Ellrott et al. 2018* as an example dataset here. These data are publicly available at *https://gdc.cancer.gov/about-data/publications/mc3-2017*

```{sh, eval=F}
# Takes some time to download, +/- 1GB
wget https://api.gdc.cancer.gov/data/1c8cfe5f-e52d-41ba-94da-f15ea1337efc -O downloads/MC3.maf
```

# Process mutation data

```{r}
# Load data
maf_MC3<- read.table("downloads/MC3.maf",header = T,stringsAsFactors = F)

# Check whether genome build was hg19 (GRCh37), as the annotation was only developed for this build.
table(maf_MC3$NCBI_Build) # all GRCh37

# Extract point mutations from data
maf_MC3<- maf_MC3[maf_MC3$Variant_Type=="SNP",]

# Extract the following variables: gene name, chromosome, start (end) position, ref allele, alt allele, sample id, variant.
maf_MC3<- maf_MC3[,c("Hugo_Symbol","Chromosome","Start_Position","End_Position","Reference_Allele","Tumor_Seq_Allele2","Tumor_Sample_Barcode","Variant_Classification")]

# Main chromosomes only
maf_MC3<- maf_MC3[maf_MC3[,"Chromosome"]%in%c(1:22,"X","Y"),]

# Add subst types
source("../scripts/functions/get_subst_types.R")
subst_type3<- get_subst_types(maf_MC3$Chromosome,maf_MC3$Start_Position,maf_MC3$"Tumor_Seq_Allele2")[["subst_type3"]]
maf_MC3<- cbind(maf_MC3,subst_type3,stringsAsFactors=F)

# Add cancer information from TCGA manifest
load("../data/TCGA_manifest.RData")
maf_MC3$Tumor_Sample_Barcode<- substr(maf_MC3$Tumor_Sample_Barcode,1,15)
maf_MC3$Cancer<- TCGA_cancer_id[maf_MC3$Tumor_Sample_Barcode,1]

# Convert to maf file to GPos object
source("../scripts/functions/convert_maf_to_GPos.R")
metadata_cols<- c("Hugo_Symbol","Variant_Classification","subst_type3","Tumor_Sample_Barcode","Cancer")
maf_GP<- Convert_maf_to_GPos.R(maf = maf_MC3,metadata_cols = metadata_cols)

# Match variant names to annotation format
maf_GP[maf_GP$Variant_Classification=="Missense_Mutation"]$Variant_Classification<- "nonsynonymous SNV"
maf_GP[maf_GP$Variant_Classification=="Silent"]$Variant_Classification<- "synonymous SNV"

# Save
saveRDS(maf_GP,file = "downloads/maf_MC3_GP.rds")
```

# Add HLA annotation to mutation data

```{r}
maf_GP<- readRDS(file = "downloads/maf_MC3_GP.rds")

# Load GRanges files containing information on all annotated regions
load("../data/annotation_gr.RData")

# Add HLA-binding annotation to maf
seqlevels(maf_GP)<- paste0("chr",unique(as.character(seqnames(maf_GP)))) # Match seqlevels first if necessary 
maf_GP_present_idx<- findOverlaps(maf_GP,annotation_gr_all,ignore.strand=T)
maf_GP_HLA_idx<- findOverlaps(maf_GP,annotation_gr_HLA_Mean,ignore.strand=T)
maf_GP$isHA<- (1:length(maf_GP))%in%queryHits(maf_GP_HLA_idx)
maf_GP$isHA[!(1:length(maf_GP))%in%queryHits(maf_GP_present_idx)]<- NA # Remove CDS position with no information
cat("HLA-binding mutations: ", 100*signif(mean(maf_GP$isHA,na.rm=T),3),"%"," \n")

# Save
saveRDS(maf_GP,file = "downloads/maf_MC3_GP_annotated.rds")
```

# HBMR metric

```{r}
# Load 
maf_GP<- readRDS(file = "downloads/maf_MC3_GP_annotated.rds")

# Calculate for cancers studied in the manuscript.
cancers<- sort(c('BRCA','COADREAD','LUAD','OV','HNSC','UCEC','LUSC','PRAD','LGG','STAD','SKCM','GBM','BLCA','LIHC','CESC','KIRP','SARC','PAAD','ESCA'))

# Calculate HBMR metric
source("../scripts/functions/calculate_HBMR.R")
subst_type_matrix_ls<- readRDS(file="../data/subst_type_matrix_ls.rds") # Expected mut per subst type
HBMR<- calculate_HBMR(sample = maf_GP$Tumor_Sample_Barcode,
                      variant = maf_GP$Variant_Classification,
                      isHA = maf_GP$isHA,
                      subst_type3 = maf_GP$subst_type3,
                      exp_mut = subst_type_matrix_ls$triNT,
                      cancer =  maf_GP$Cancer)[cancers,]
HBMR<- HBMR[order(HBMR[,"HBMR_norm"]),] # Sort by HBMRnorm

# Plot
bp<- barplot(HBMR[,"HBMR_norm"],las=2,ylab="Normalized HBMR",ylim=c(0,1.3),main="HBMR metric")
arrows(bp,HBMR[,"HBMR_norm"],bp,HBMR[,"HBMR_norm_ciup"],lwd = 1.5, angle = 90,code = 2, length = 0.03,xpd=NA)    
abline(h=1,lty=2)
```

Barplot with normalized HBMR and 95% CI for each cancer.

# dNHLA/dNnonHLA metric

This metric was developed to compare observed and expected HLA affinities for mutated peptides from each individual sample, considering its HLA genotype. This requires HLA genotyping of samples and prediction of mutated peptide affinities. Because HLA genotypes are derived from protected TCGA data, we are not allowed to make them available. For more information on how to use this metric see  *scripts/manuscript_Rsim_HLASpecific.R* or contact the corresponding author.


